src/compiler.rkt

```diff
@@ -2,31 +2,106 @@
+;;                    | <operation> - const
 ;; ```
 ;; where <index> refers to a previous virtual register.
 ;; Must also provide the input variables for the program(s)
 ;; as well as the indices of the roots to extract.
-(define (make-progs-interpreter vars ivec roots)
+;; name ::= 'fl or 'ival
+(define (make-progs-interpreter name vars ivec roots)
   (define vreg-count (+ (length vars) (vector-length ivec)))
   (define vregs (make-vector vreg-count))
-  (λ args
-    (for ([arg (in-list args)] [n (in-naturals)])
-      (vector-set! vregs n arg))
-    (for ([instr (in-vector ivec)] [n (in-naturals (length vars))])
-      (define srcs
-        (for/list ([idx (in-list (cdr instr))])
-          (vector-ref vregs idx)))
-      (vector-set! vregs n (apply (car instr) srcs)))
-    (for/list ([root (in-list roots)])
-      (vector-ref vregs root))))
+
+  (define (unbox-prec x)
+    (if (box? x)
+        (unbox x)
+        0))
+
+  (define (tuning-filter instr)
+    (if (vector? (car instr))
+        (if (member
+             (vector-ref (car instr) 0)
+             (list ival-sin ival-cos ival-tan))
+            #t
+            #f)
+        #f))
+  
+  (define tuning-ivec
+    (if (equal? name 'ival)
+        (vector-filter tuning-filter ivec)
+        '()))
+  
+  (if (equal? name 'ival)
+    (λ args
+      ;; remove all the exponent values we assigned previously when a new point comes
+      (when (equal? (bf-precision) (*starting-prec*))
+        (for ([instr (in-vector tuning-ivec)])
+          (set-box! (vector-ref (car instr) 2) 0)))
+    
+      (for ([arg (in-list args)] [n (in-naturals)])
+        (vector-set! vregs n arg))
+      (for ([instr (in-vector ivec)] [n (in-naturals (length vars))])
+        (define srcs
+          (for/list ([idx (in-list (cdr instr))])
+            (vector-ref vregs idx)))
+
+        (match (car instr)
+          [(vector op extra-precision)
+           (let ([extra-prec (unbox-prec extra-precision)])
+             (if (zero? extra-prec)
+                 (vector-set! vregs n (apply op srcs))
+                 (parameterize ([bf-precision (define-precision extra-prec)])
+                   (vector-set! vregs n (apply op srcs)))))]
+          
+          [(vector op extra-precision exponents-checkpoint)  ; current op is sin/cos/tan
+           (let ([extra-prec (unbox-prec extra-precision)])
+             (if (zero? extra-prec)
+                 (vector-set! vregs n (apply op srcs))
+                 (parameterize ([bf-precision (define-precision extra-prec)])
+                   (vector-set! vregs n (apply op srcs))))
+             (set-box! exponents-checkpoint ; Save exponents with the passed precision for the next run
+                       (max 0
+                            (+ extra-prec
+                               (max (true-exponent (ival-lo (car srcs)))
+                                    (true-exponent (ival-hi (car srcs))))))))]
```

Comment:
If we start supporting other tunable operations, we'll need to make this thing specializable for different operators, right?

Reply:
Yes, on my another branch `artem-minus-and-trig-tuning` I have different formulas for tuning minus and trigs, and there it is kinda messy. The logic takes too much space inside compiler and ideally I think it should somewhere in another file.